martin/src/cog/source.rs

```diff
@@ -1,68 +1,136 @@
     pub fn new(id: String, path: PathBuf) -> FileResult<Self> {
         let tileinfo = TileInfo::new(Format::Png, martin_tile_utils::Encoding::Uncompressed);
-        let meta = get_meta(&path)?;
+        let tif_file =
+            File::open(&path).map_err(|e: std::io::Error| FileError::IoError(e, path.clone()))?;
+        let mut decoder = Decoder::new(tif_file)
+            .map_err(|e| CogError::InvalidTiffFile(e, path.clone()))?
+            .with_limits(tiff::decoder::Limits::unlimited());
+        let model = ModelInfo::decode(&mut decoder, &path);
+        verify_requirements(&mut decoder, &model, &path.clone())?;
+        let nodata: Option<f64> = if let Ok(no_data) = decoder.get_tag_ascii_string(GdalNodata) {
+            no_data.parse().ok()
+        } else {
+            None
+        };
+        let origin = get_origin(
+            model.tie_points.as_deref(),
+            model.transformation.as_deref(),
+            &path,
+        )?;
+        let (full_width_pixel, full_length_pixel) = dimensions_in_pixel(&mut decoder, &path, 0)?;
+        let (full_width, full_length) = dimensions_in_model(
+            &mut decoder,
+            &path,
+            0,
+            model.pixel_scale.as_deref(),
+            model.transformation.as_deref(),
+        )?;
+        let extent = get_extent(
+            &origin,
+            model.transformation.as_deref(),
+            (full_width_pixel, full_length_pixel),
+            (full_width, full_length),
+        );
+
+        let mut images = vec![];
+        let mut ifd_index = 0;
+
+        loop {
+            let is_image = decoder
+                .get_tag_u32(Tag::NewSubfileType)
+                .map_or_else(|_| true, |v| v & 4 != 4); // see https://www.verypdf.com/document/tiff6/pg_0036.htm
+            if is_image {
+                // TODO: We should not ignore mask in the next PRs
+                images.push(get_image(&mut decoder, &path, ifd_index)?);
+            } else {
+                warn!(
+                    "A subfile of {} is ignored in the tiff file as Martin currently does not support mask subfile in tiff. IFD={ifd_index}",
+                    path.display(),
+                );
+            }
+
+            ifd_index += 1;
+
+            let next_res = decoder.seek_to_image(ifd_index);
+            if next_res.is_err() {
+                // TODO: add warn!() here
+                break;
+            }
+        }
+        let min_zoom = 0;
+        let max_zoom = (images.len() - 1) as u8;
+        let images: HashMap<u8, Image> = images
+            .iter()
+            .enumerate()
+            .map(|(idx, image)| {
+                // FIXME: explain why it is OK to use IFD index as zoom level,
+                //        and why it would not exceed max_zoom.
+                //        This looks like a bug for some reason.
+                let zoom = max_zoom.saturating_sub(idx as u8);
+                (zoom, image.clone())
```

Comment:
this is inefficient - you clone everything

```suggestion
            .into_iter()
            .enumerate()
            .map(|(idx, image)| {
                // FIXME: explain why it is OK to use IFD index as zoom level,
                //        and why it would not exceed max_zoom.
                //        This looks like a bug for some reason.
                let zoom = max_zoom.saturating_sub(idx as u8);
                (zoom, image)
```